lib/membrane/core/element/caps_controller.ex

```diff
@@ -50,26 +50,36 @@ defmodule Membrane.Core.Element.CapsController do
   @spec exec_handle_caps(Pad.ref_t(), Caps.t(), params :: map, State.t()) :: State.t()
   def exec_handle_caps(pad_ref, caps, params \\ %{}, state) do
     require CallbackContext.Caps
-    %{accepted_caps: accepted_caps} = PadModel.get_data!(state, pad_ref)
+
+    %{parents_with_pads: parents_with_pads, name: pad_name} = PadModel.get_data!(state, pad_ref)
+
     context = &CallbackContext.Caps.from_state(&1, pad: pad_ref)
 
-    if Caps.Matcher.match?(accepted_caps, caps) do
-      state =
-        CallbackHandler.exec_and_handle_callback(
-          :handle_caps,
-          ActionHandler,
-          %{context: context} |> Map.merge(params),
-          [pad_ref, caps],
-          state
-        )
+    :ok = ensure_caps_match(:input, [{state.module, pad_name} | parents_with_pads], caps)
 
-      PadModel.set_data!(state, pad_ref, :caps, caps)
-    else
-      raise """
-      Received caps: #{inspect(caps)} that are not specified in def_input_pad
-      for pad #{inspect(pad_ref)}. Specs of accepted caps are:
-      #{inspect(accepted_caps, pretty: true)}
-      """
+    state =
+      CallbackHandler.exec_and_handle_callback(
+        :handle_caps,
+        ActionHandler,
+        %{context: context} |> Map.merge(params),
+        [pad_ref, caps],
+        state
+      )
+
+    PadModel.set_data!(state, pad_ref, :caps, caps)
+  end
+
+  @spec ensure_caps_match(Pad.direction_t(), [{module(), Pad.name_t()}], Caps.t()) :: :ok
+  def ensure_caps_match(direction, modules_with_pads, caps) do
+    for {module, pad_name} <- modules_with_pads do
+      if not module.caps_match?(pad_name, caps) do
```

Comment:
It's a unique occasion to use unless :P

```suggestion
      unless module.caps_match?(pad_name, caps) do
```